sql,tree: differentiate UNIQUE and UNIQUE INDEX in CREATE TABLE

ajwerner opened the pull request:
While we treated the cases of `ALTER TABLE ... ADD CONSTRAINT UNIQUE` from
`CREATE UNIQUE INDEX` we didn't treat the two clauses differently in the
context of `CREATE TABLE`. While functionally they are largely identical concepts,
they have different semantic meanings (you can't use drop constraint on the index
or drop index on the constraint). We also didn't differentiate these concepts for display.

The root of this problem is twofold:

1) Unique indexes and unique constraints always displayed indentically in `SHOW CREATE TABLE` even though some other catalog fields treated them distinctly.
2) `UNIQUE` and `UNIQUE INDEX` clauses were treated identically in `CREATE TABLE` statements.

```
root@localhost:26257/test> CREATE TABLE t (i INT PRIMARY KEY, j INT, UNIQUE (j), UNIQUE INDEX (j));
CREATE TABLE
```

Before:
```
root@localhost:26257/test> SHOW CREATE TABLE t;
  table_name |               create_statement
-------------+------------------------------------------------
  t          | CREATE TABLE public.t (
             |     i INT8 NOT NULL,
             |     j INT8 NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (i ASC),
             |     UNIQUE INDEX t_j_key (j ASC),
             |     UNIQUE INDEX t_j_key1 (j ASC),
             |     FAMILY "primary" (i, j)
             | )
```

After:
```
demo@127.0.0.1:55769/movr> SHOW CREATE TABLE t;
  table_name |               create_statement
-------------+------------------------------------------------
  t          | CREATE TABLE public.t (
             |     i INT8 NOT NULL,
             |     j INT8 NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (i ASC),
             |     UNIQUE INDEX t_j_key1 (j ASC),
             |     FAMILY "primary" (i, j),
             |     CONSTRAINT t_j_key UNIQUE (j)
             | )
```

Release note (bug fix): UNIQUE INDEX clauses in a CREATE TABLE statement
are no longer treated as CONSTRAINT UNIQUE clauses.

Release note (bug fix): UNIQUE constraints were displayed as UNIQUE INDEX
entries in SHOW CREATE TABLE.

ajwerner committed:
sql,tree: differentiate UNIQUE and UNIQUE INDEX in CREATE TABLE

While we treated the cases of `ALTER TABLE ... ADD CONSTRAINT UNIQUE` from
`CREATE UNIQUE INDEX` we didn't treat the two clauses differently in the
context of `CREATE TABLE`. While functionally they are indentical concepts,
they do display differently and have different semantic meanings (you can't
use drop constraint on the index or drop index on the constraint).

Release note (bug fix): UNIQUE INDEX clauses in a CREATE TABLE statement
are no longer treated as CONSTRAINT UNIQUE clauses.

ajwerner committed:
sql: SHOW CREATE TABLE treated unique constraints like unique indexes

```
root@localhost:26257/test> CREATE TABLE t (i INT PRIMARY KEY, j INT);
CREATE TABLE
root@localhost:26257/test> ALTER TABLE t ADD CONSTRAINT c UNIQUE (j);
ALTER TABLE
root@localhost:26257/test> CREATE UNIQUE INDEX idx ON t (j);
CREATE INDEX
```

Before:
```
root@localhost:26257/test> SHOW CREATE TABLE t;
  table_name |               create_statement
-------------+------------------------------------------------
  t          | CREATE TABLE public.t (
             |     i INT8 NOT NULL,
             |     j INT8 NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (i ASC),
             |     UNIQUE INDEX c (j ASC),
             |     UNIQUE INDEX idx (j ASC),
             |     FAMILY "primary" (i, j)
             | )
```

After:
```
demo@127.0.0.1:55792/movr> SHOW CREATE TABLE t;
  table_name |               create_statement
-------------+------------------------------------------------
  t          | CREATE TABLE public.t (
             |     i INT8 NOT NULL,
             |     j INT8 NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (i ASC),
             |     UNIQUE INDEX idx (j ASC),
             |     FAMILY "primary" (i, j),
             |     CONSTRAINT c UNIQUE (j)
             | )
```

Release note (bug fix): UNIQUE constraints were displayed as UNIQUE INDEX
entries in SHOW CREATE TABLE.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65825)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Nice cleanup!

> Reviewed 1 of 2 files at r1, 21 of 21 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-:-MaoYtkw07eZWiObUF8R:bt3oa4v)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[pkg/sql/show_create_clauses.go, line 604 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MaoULI5Ct-Cqeb_Yh0y:-MaoULI5Ct-Cqeb_Yh0z:b7jj9rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f10d2b64bbb2ef7f5d3a082fbd0751d95480c4ca/pkg/sql/show_create_clauses.go#L604)):*
> ```Go
> 		}
> 	}
> 	for _, idx := range desc.AllIndexes() {
> ```

[nit] Maybe we want to show them before `UNIQUE WITHOUT INDEX` (since they are more "vanilla")? Also, I'd add a comment along the lines of `// Show unique constraints that are implemented by implicitly created unique indexes`.
___
*[pkg/sql/logictest/testdata/logic_test/show_create_all_tables, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MaoZ-miBtTnbupBMusj:-MaoZ-miBtTnbupBMusk:blnvf8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2e283478266494205fe95de98270e260e08d26/pkg/sql/logictest/testdata/logic_test/show_create_all_tables#L59)):*
> ```
> );
> CREATE TABLE public.full_test (
>                                                                                                                                                                      x INT8 NULL,
> ```

What's with the extra blank lines? Maybe a bug in `-rewrite`?
___
*[pkg/sql/sem/tree/create.go, line 885 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MaoYIEz3swQHu4XVAvE:-MaoYIEz3swQHu4XVAvF:bb0ivbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2e283478266494205fe95de98270e260e08d26/pkg/sql/sem/tree/create.go#L885)):*
> ```Go
> 	PrimaryKey    bool
> 	WithoutIndex  bool
> 	ExplicitIndex bool
> ```

[nit] Maybe add a comment for the field (mentioning CONSTRAINT vs UNIQUE INDEX syntax). And that at most one of PrimaryKey, WithoutIndex, ExplicitIndex can be true.


<!-- Sent from Reviewable.io -->

ajwerner committed:
sql: do not show CHECK or UNIQUE WITHOUT INDEX constraints currently being added

This is potentially controversial but it will make things more akin to other
behaviors. It retains `NOT VALID` for constraints created as `NOT VALID` but
it does not display constraint which are in the process of being validated.

We currently do not display indexes (unique or otherwise) or foreign key
constraints which are in the process of being added. I think we should
pick one and stick to it.

Release note (sql change): SHOW CREATE TABLE will no longer show CHECK or
UNIQUE WITHOUT INDEX constraints which are in the process of being added.
In earlier versions, they would appear with NOT VALID. Now NOT VALID will
only appear for constraints which were created that way.

rytaft reviewed (commented):
Nice change -- this will get us closer to PG compatibility, since Postgres distinguishes between unique constraints and unique indexes. 

This issue seems relevant, since it's related to the fact that we weren't distinguishing between unique constraints and unique indexes before: https://github.com/cockroachdb/cockroach/issues/42840. (not that you need to fix it with this PR, just wanted to surface it)

> Reviewed 2 of 2 files at r1, 21 of 21 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-:-MapJA_G6-afkaloTPJR:b-2w377s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @mgartner, and @RaduBerinde)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_index, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MapGFxe4Vl4oKVefoKd:-MapGFxe4Vl4oKVefoKe:be9dq2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2e283478266494205fe95de98270e260e08d26/pkg/ccl/logictestccl/testdata/logic_test/partitioning_index#L50)):*
> ```
>      CONSTRAINT "primary" PRIMARY KEY (a ASC),
>      FAMILY "primary" (a, b),
>      CONSTRAINT ok2_b_key UNIQUE (b)
> ```

the partition by list isn't shown now
___
*[pkg/sql/show_create_clauses.go, line 604 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MaoULI5Ct-Cqeb_Yh0y:-MapGe30AQDNFsxQRsyJ:bcxjzav) ([raw file](https://github.com/cockroachdb/cockroach/blob/f10d2b64bbb2ef7f5d3a082fbd0751d95480c4ca/pkg/sql/show_create_clauses.go#L604)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Maybe we want to show them before `UNIQUE WITHOUT INDEX` (since they are more "vanilla")? Also, I'd add a comment along the lines of `// Show unique constraints that are implemented by implicitly created unique indexes`.
</blockquote></details>

I'd also show them before the families
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 279 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65825#-MapGpn47JhMcC-gr_u9:-MapGpn47JhMcC-gr_uA:b-cgrigk) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2e283478266494205fe95de98270e260e08d26/pkg/sql/logictest/testdata/logic_test/create_table#L279)):*
> ```
>             INVERTED INDEX like_table_j_idx (j),
>             FAMILY "primary" (a, b, c, h, j, k),
>             CONSTRAINT foo UNIQUE (b, c)
> ```

now the DESC marker is gone from b


<!-- Sent from Reviewable.io -->

ajwerner commented:
Thanks for the early looks! There's a good bit more I want to do here. I'll mark it as `Ready for review` when it's in good shape.

1) Make the behavior consistent for partially added things.
   * See [internal thread](https://cockroachlabs.slack.com/archives/C04U1BTF8/p1622220436082500)
2) Change the parser to disallow a bunch of the index parameters on unique constraints 
   * We'll need to allow WHERE clauses on UNIQUE WITHOUT INDEX because we can't take
      that feature away. It's fine because it's a cockroach extension.
   * Other modifiers will be removed from `UNIQUE` constraints for the most part
3) (follows from 2) Implement a migration and repair logic to turn all of the unique constraints which use index features into unique indexes explicitly. This includes:
   * non-standard column directions,
   * partition clauses
   * interleave clauses
   * storing (technically in postgres you can have an `INCLUDE CLAUSE` but we don't support that syntax). When we add it, I'm fine supporting it in both cases. 
4) Fix some other bugs revealed along the way. 
   * For example, we disallow `NOT VALID` on `UNIQUE` in a `CREATE TABLE` but we ignore it in when adding a constraint via `ALTER TABLE`. https://github.com/cockroachdb/cockroach/blob/c1f753457cf8e73e417c875fcf52a06d4ac3e4ce/pkg/sql/parser/sql.y#L6206
   * There's some other issues with adding multiple constraints in a single `ALTER TABLE` I'm chewing through.

ajwerner committed:
sql: fix a bug which prevented self referencing NOT VALID foreign keys

This bug is due to both using the same object in descs.Txn (sort of) and the
fact the fact that MakeMutationComplete does not remove the mutation any
longer. I have a feeling, but have not checked, that we lost this back-
reference in older versions.

Release note (bug fix): Fixed a bug which prevented adding self-referencing
FOREIGN KEY constraints in the NOT VALID state.

ajwerner committed:
sql: fix NOT VALID validation in parser

This was the behavior in CREATE TABLE but not in ALTER TABLE. The impact was
that the parser would allow such clauses but ignore them. Secondarily, we
permitted (and supported) using NOT VALID for UNIQUE WITHOUT INDEX in
ALTER TABLE but not in CREATE TABLE. I've added support for it for symmetry.
I don't know that there's ever a reason one would want to do that though.

UNIQUE WITHOUT INDEX constraints may still be created in a NOT VALID state.

Release note (bug fix): An error is now reported when using NOT VALID with
a UNIQUE constraint in ALTER TABLE. Previously the clause was silently
ignored.

ajwerner committed:
sql,parser: disallow index parameters on UNIQUE constraints

Before this commit we would allow, in the grammar, a variety of options like
interleaving, partitioning, and storing as well as attributes on columns
related to sorting. None of these are valid in postgres nor do they have much
to do with the constraint. Instead, we remove all of these from the grammar and
hint to users that `UNIQUE INDEX` is what they are looking for.

This is a backwards incompatible change; prior to this change, one could use
the syntax of a unique constraint to construct a unique index. We accept this
change because it drives closer to postgresql compatibility.

Furthermore, because of bugs in how we display these constraints, it was
very difficult to determine whether something was a unique constraint or
a unique index.

Release note (backward-incompatible change): The use of PARTITION BY,
INTERLEAVE, and STORING clauses as well as ordering parameters on columns
in UNIQUE constraint definitions is no longer supported; instead, users
are directed towards UNIQUE INDEX which supports all of these features.

ajwerner committed:
sql: properly generate UNIQUE INDEX names with idx suffix

Postgres names the indexes due to unique constraints (as well as the
constraint) with the `key` suffix but names unique indexes like regular
indexes, with the `idx` suffix.

Release note (sql change): Indexes which are automatically named using
a CREATE UNIQUE INDEX statement or a UNIQUE INDEX clause in a CREATE
TABLE statement now have the suffix idx instead of key which is used
for UNIQUE constraints. This matches the postgres behavior.

mgartner commented:
@ajwerner Do you think we should try to get this into 21.2? I ask not to pressure you, but because there's some related issues that have come up which would be solved by this. There's other temporary solutions that I can work on if we think this PR is too big of a lift for 21.2. I could also take over where you left off.

* https://github.com/cockroachdb/cockroach/issues/67234
* This doesn't really need to be solved, but would be affected by: https://github.com/cockroachdb/cockroach/pull/67197/files#diff-931c3f5a978f27bc78dffa628d45152d0c8c6f8be4328c5c5b2e760c28f2b32aR132-R136

ajwerner commented:
> @ajwerner Do you think we should try to get this into 21.2? I ask not to pressure you, but because there's some related issues that have come up which would be solved by this. There's other temporary solutions that I can work on if we think this PR is too big of a lift for 21.2. 

Thanks for the ping and I'm very sorry I let this sit for so long. Increasingly I'm afraid of breaking compatibility with ourselves, even in cases where we handled it poorly in the past. I'd like @vy-ton to get involved here as we do some decision making. Maybe the right approach here is to treat all of the uses of index features in a constraint clause as valid but as syntactically implying an `INDEX` as opposed to a constraint. More concretely, revert the parser changes but retain the other changes to convert those constraints to indexes. Then during `CREATE TABLE` and `ALTER TABLE ... ADD CONSTRAINT`, actually construct an `ExplicitlyCreated` index, perhaps with a notice. That would give us some latitude to disallow it one day but would be less disruptive if anybody is depending on it. 

> I could also take over where you left off.

As far as getting this done, is this something you feel like you have resources to tackle?

mgartner commented:
> As far as getting this done, is this something you feel like you have resources to tackle?

Probably not before the stability period. I'll look for quick workarounds for the minor issues I linked to for now, and hopefully we can make some progress on this in 22.1.
